channeldb/invoices.go

```diff
@@ -505,48 +511,68 @@ func (d *DB) LookupInvoice(paymentHash [32]byte) (Invoice, error) {
 	return invoice, nil
 }
 
-// FetchAllInvoices returns all invoices currently stored within the database.
-// If the pendingOnly param is true, then only unsettled invoices will be
-// returned, skipping all invoices that are fully settled.
+// FetchAllInvoices returns all invoices and their payment hashes
+// currently stored within the database. If the pendingOnly param is true, then
+// only unsettled invoices and their payment hashes will be returned, skipping
+// all invoices that are fully settled or canceled.
 func (d *DB) FetchAllInvoices(pendingOnly bool) ([]Invoice, error) {
-	var invoices []Invoice
+	var fetchedInvoices []Invoice
+	var fetchedInvoiceNums []uint32
 
 	err := d.View(func(tx *bbolt.Tx) error {
-		invoiceB := tx.Bucket(invoiceBucket)
-		if invoiceB == nil {
+		invoices := tx.Bucket(invoiceBucket)
+		if invoices == nil {
 			return ErrNoInvoicesCreated
 		}
 
-		// Iterate through the entire key space of the top-level
-		// invoice bucket. If key with a non-nil value stores the next
-		// invoice ID which maps to the corresponding invoice.
-		return invoiceB.ForEach(func(k, v []byte) error {
-			if v == nil {
+		invoiceIndex := invoices.Bucket(invoiceIndexBucket)
+		if invoiceIndex == nil {
+			return ErrNoInvoicesCreated
+		}
+
+		return invoiceIndex.ForEach(func(paymentHash, invoiceNum []byte) error {
+			// Due to legacy reasons, we need to check if the payment hash
+			// equals to the numInvoicesKey as that will generate a non valid
+			// not found error.
+			if invoiceNum == nil || reflect.DeepEqual(paymentHash, numInvoicesKey) {
 				return nil
 			}
 
-			invoiceReader := bytes.NewReader(v)
-			invoice, err := deserializeInvoice(invoiceReader)
+			invoice, err := fetchInvoice(invoiceNum, invoices)
 			if err != nil {
 				return err
 			}
 
 			if pendingOnly &&
-				invoice.Terms.State == ContractSettled {
+				(invoice.Terms.State == ContractSettled ||
+					invoice.Terms.State == ContractCanceled) {
 
 				return nil
 			}
 
-			invoices = append(invoices, invoice)
+			// Fill the optional Invoice.PaymentHash with the one from the index.
+			invoice.PaymentHash = make([]byte, lntypes.HashSize)
+			copy(invoice.PaymentHash, paymentHash[:lntypes.HashSize])
+			fetchedInvoices = append(fetchedInvoices, invoice)
+
+			// Save the invoice num as we'll need it at the and to restore order.
+			fetchedInvoiceNums = append(
+				fetchedInvoiceNums, byteOrder.Uint32(invoiceNum))
```

Comment:
Close paren should being on new line below.

Reply:
done